Implement compression for raw data files

andrewmogan opened the pull request:
This PR introduces a `compression_level` member into `HDF5RawDataFile` which sets the gzip compression level using HighFive's `Deflate` method. The default value is 0 (no compression). If set to a non-zero value, each dataset will be compressed using `raw_data_size_bytes` as the chunk size. For a simple, five second run, this reduces the total file size by almost a factor of 3. I can also confirm interactively using `h5stat -S` that the total size of all datasets is reduced when `compression_level` is set. 

In order to preserve the functionality of `HDF5LIBS_TestDumpRecord`, the buffer size allocated by `data_size` needed to be adjusted to not use `getStorageSize()`, which does not account for possible compression. `data_size` is now set as
`size_t data_size = data_set.getSpace().getElementCount() * sizeof(char);`, where `sizeof(char)` is basically a placeholder that can be adjusted in the future if necessary. 

The largest change here is to the unit tests. The functionality is the same, but for the sake of running each test with and without compression, I created a `FileWriteFixture` to avoid copy-pasting each test. 

This PR also relies on parallel PRs to [appmodel](https://github.com/DUNE-DAQ/appmodel/pull/164), [daqsystemtest](https://github.com/DUNE-DAQ/daqsystemtest/pull/171), and [dfmodules](https://github.com/DUNE-DAQ/dfmodules/pull/396). 

There are still some to-do's (listed below), but future, related PRs will be easier once these initial changes are merged. 

Still to-do's:
- Make chunk size configurable
- Compress TP stream files in addition to raw data files (I'm not yet sure why this isn't already happening)

andrewmogan committed:
Add simple compression for raw data file

andrewmogan committed:
Decrease compression level

andrewmogan committed:
Add compression_level to unit tests

andrewmogan committed:
Make compression_level configurable

andrewmogan committed:
Account for compression_level in test utilities

andrewmogan committed:
Reorder variables and remove debug print statements

andrewmogan committed:
Handle compression in membuffer size

The previous method using getStorageSize() only allocated membuffer
based on the size of the stored data, which breaks file reading in the
case where the datasets are compressed. Using the total number of
elements multiplied by the size of the type (char) fixes this.

Also, removed a bunch of logging statements that are no longer
necessary.

andrewmogan committed:
Add compression_level information

andrewmogan committed:
Factor get and delete methods into new FileUtils header

andrewmogan committed:
Use specialized template for create_file_layout_params

andrewmogan committed:
Move reused code to new HDF5TestUtils.hpp

andrewmogan committed:
Move HDF5TestUtils.hpp to new test subdirectory

andrewmogan committed:
Update include paths for HDF5TestUtils

andrewmogan committed:
Revert specialized templates

andrewmogan committed:
Fix function calls following reversion of template functions

andrewmogan committed:
Remove const-ness of compression_level

andrewmogan committed:
Implement FileWriteFixture to reduce code duplication

andrewmogan committed:
Implement total dataset size parameter to distinguish from size on disk

andrewmogan committed:
Merge branch 'develop' of https://github.com/DUNE-DAQ/hdf5libs into amogan/hdf5_compression_develop

andrewmogan committed:
Rename variable which tracks dataset size on disk

andrewmogan committed:
Check compression level in ReadFileAttributes test

andrewmogan committed:
Move remaining test functions to test fixture

andrewmogan committed:
Add FileWriteFixture to TriggerRecord unittest

andrewmogan committed:
Sync FileWriteFixture functionality between TriggerRecord and TimeSlice tests

andrewmogan committed:
Move all tests to FileWriteFixture struct

This includes extending the fixture to account for large trigger record
numbers. Also, fix some compiler errors.

andrewmogan committed:
Add checks for compressed file size

andrewmogan committed:
Remove total dataset size parameter and some debug statements

andrewmogan committed:
Correct compression_level type

andrewmogan committed:
Remove total dataset size and attribute printing

andrewmogan committed:
Add compression_level as a variable rather than hard-coding

andrewmogan committed:
Fix seq_num type

bieryAtFnal commented:
Sorry for taking so long to look at these changes...  

I'll summarize my current comments here, even though there are PRs in other repositories...

1. These changes are not correctly handling the maximum size for files when compression is enabled.  I suspect that this is because the size of the uncompressed data is being added to the 'recorded size' instead of the compressed size.  Apologies if I contributed to confusion on this point in our earlier discussion.  I can send sample instructions for reproducing the problem, if you want.
2. I think that it would be helpful to add a "`<attr name="compression_level" type="u8" val="0"/>`" line to the `default_tp_store_conf` entry in `daqsystemtest/config/daqsystemtest/moduleconfs.data.xml` so that users can see how to enable compression for TP stream files.
3. I wonder if the changes to `HDF5LIBS_TestDumpRecord.cpp` can be made backward-compatible by checking if the compression_level attribute is present in the data file before fetching it.  (Backward-compatible in this case means being able to run `HDF5LIBS_TestDumpRecord` on files produced before compression was implemented.)
   1. Also, the compression level printout is empty when I run `HDF5LIBS_TestDumpRecord` on a file that has compression enabled.
5. In the `HDF5WriteReadTriggerRecord_test` (unit test) `FileWriteFixture`, the `trigger_record_number_offset` is not used in the calculation of the `trigger_number` when it is non-zero.
   1. Also, `fl_pars` is re-defined and re-assigned in the `FileWriteFixture` constructor.
7. Is there a place where the supported values of the compression level are listed?

andrewmogan committed:
Check for compression_level existence before fetching it

bieryAtFnal commented:
Here are sample instructions for the first item in my previous post...

```
DATE_PREFIX=`date '+%d%b'`
TIME_SUFFIX=`date '+%H%M'`

source /cvmfs/dunedaq.opensciencegrid.org/setup_dunedaq.sh
setup_dbt latest_v5
dbt-create -n NFD_DEV_250116_A9 ${DATE_PREFIX}FDDevTest_${TIME_SUFFIX}
cd ${DATE_PREFIX}FDDevTest_${TIME_SUFFIX}/sourcecode

git clone https://github.com/DUNE-DAQ/appmodel.git -b amogan/hdf5_compression_develop
git clone https://github.com/DUNE-DAQ/dfmodules.git -b amogan/hdf5_compression_develop
git clone https://github.com/DUNE-DAQ/daqsystemtest.git -b amogan/hdf5_compression_develop
git clone https://github.com/DUNE-DAQ/hdf5libs.git -b amogan/hdf5_compression_develop
cd ..

dbt-workarea-env
dbt-build -j 12
dbt-workarea-env

daqconf_set_connectivity_service_port local-1x1-config config/daqsystemtest/example-configs.data.xml

sed -i 's,<attr name="compression_level" type="u8" val="0"/>,<attr name="compression_level" type="u8" val="1"/>,' sourcecode/daqsystemtest/config/daqsystemtest/moduleconfs.data.xml

sed -i 's,<attr name="max_file_size" type="u64" val="4294967296"/>,<attr name="max_file_size" type="u64" val="4194304"/>,' sourcecode/daqsystemtest/config/daqsystemtest/moduleconfs.data.xml

mkdir rundir
cd rundir

drunc-unified-shell ssh-standalone config/daqsystemtest/example-configs.data.xml local-1x1-config boot --no-override-logs wait 5 conf wait 3 start --run-number 101 enable-triggers wait 40 disable-triggers drain-dataflow stop-trigger-sources stop scrap terminate

echo
echo
ls -alF *.hdf5
echo
echo "Note that the file sizes are much less than the requested max_file_size of 4 MiB..."
```

andrewmogan committed:
Change uint8_t to unsigned

This was causing compression_level to print as blank even when it was
set correctly.

andrewmogan committed:
Implement get_attribute_if_exists and logical_size

andrewmogan committed:
Modify do_write to return uncompressed size.

This is in addition to the recorded size. Also, add a bunch of print
statements for debugging.

andrewmogan committed:
Only use compression if compression_level > 0

This is because enabling compression, even at level 0, changes the
storage structure from contiguous to chunked. Given how many datasets
we're writing, this massively increases the metadata overhead. This way,
compression_level 0 does not introduce unnecessary metadata.

andrewmogan committed:
Rename logical size to uncompressed size

andrewmogan committed:
Finish renaming logical size to uncompressed size

andrewmogan committed:
Remove superfluous print statement

andrewmogan committed:
Use getFileSize() to write the total file size as an attribute

andrewmogan committed:
Remove redundant fl_pars definition

andrewmogan committed:
Remove some debug prints

andrewmogan committed:
Revert uncompressed_file_size since it's no longer necessary

andrewmogan committed:
Check for compression on raw data size

Previously, we were comparing to the total file size. But enabling
compression, even at level 0, can cause a large increase in metadata
overhead, making the "compressed" file actually larger than the
uncompressed one. But we're only concerned with the raw data size being
compressed.

andrewmogan committed:
Fix an issue in trigger_number setting

andrewmogan commented:
I believe all comments are now addressed. Individual notes below:

1. The maximum file size issue has been resolved by using the `getFileSize()` method (only in HighFive v2.8 or later, included in externals v2.2). The total size is now written as an additional file attribute, separate from `recorded_size`. 
1a) Note also that I modified `m_recorded_size` to track `data_set.getStorageSize()` rather than `raw_data_size_bytes`. The number written as `recorded_size` now exactly matches the "Raw data size" number output by `h5stat -S file.hdf5`. 
1b) Also, I adjusted compressed file size check in the unit tests to check that the raw data size is compressed rather than the total file size. This is also due to the large metadata overhead introduced by enabling compression on small files with few datasets.
2. Done
3. The issue where `compression_level` wasn't printing is resolved. When run on older hdf5 files, that number will default to 0 rather than not printing at all.
3a) Additionally, compression is now only enabled if `compression_level` is greater than 0. This is because, even at compression level 0, enabling compression causes a large metadata overhead due to switching from contiguous to chunked storage. 
4. The trigger number counting logic is now simplified and appears to work consistently. Also, `fl_pars` is not re-defined. Good catches.
5. Yes, the compression level supported values are listed in the README, but I'm happy to put them in other places to improve visibility.

andrewmogan committed:
Write uncompressed raw data size as attribute

andrewmogan committed:
Merge branch 'develop' of https://github.com/DUNE-DAQ/hdf5libs into amogan/hdf5_compression_develop

andrewmogan committed:
Add support for uncompressed raw data size attribute

andrewmogan committed:
Replace cout with TLOG_DEBUG

andrewmogan committed:
Reintroduce initialization of m_recorded_size in read-only constructor (no idea why it was gone)

andrewmogan committed:
Spring cleaning

andrewmogan committed:
Merge branch 'develop' of https://github.com/DUNE-DAQ/hdf5libs into amogan/hdf5_compression_develop

andrewmogan committed:
Correctly write total_file_size attribute for reading

andrewmogan committed:
Update m_total_file_size during do_write

This is so that the current file size can be read in the HDF5DataStore
class and send that to the opmon.

andrewmogan committed:
Bump file layout version

andrewmogan committed:
Note additional file attributes in README

andrewmogan committed:
Add/update file version notes

andrewmogan committed:
Fix reference to V3 file layout notes

Commit:
Merge remote-tracking branch 'origin/develop' into amogan/hdf5_compression_develop

bieryAtFnal reviewed (approved):
LGTM.  We have verified that regression tests continue to work fine, and we've confirmed that max raw data file size limits work.  Also, the new HDF5 file Attributes look good, and the updated OpMon metrics look good.
